source/client-side-encryption/client-side-encryption.md

```diff
@@ -0,0 +1,2475 @@
+   1. If there is a `cachedAzureAccessToken` AND the duration until `azureAccessTokenExpireTime` is greater than one
+      minute, insert `cachedAzureAccessToken` as the `azure` property on $P$.
+   2. Otherwise:
+      1. Let $t_0$ be the current time.
+      2. Attempt to obtain an Azure VM Managed Identity Access Token $T$ as detailed in
+         [Obtaining an Access Token for Azure Key Vault](#obtaining-an-access-token-for-azure-key-vault).
+      3. If a token $T$ with expire duration $d\_{exp}$ were obtained successfully, create a new
+         [AzureAccessToken](#AzureAccessToken) object with $T$ as the `accessToken` property. Insert that
+         [AzureAccessToken](#AzureAccessToken) object into $P$ as the `azure` property. Record the generated
+         [AzureAccessToken](#AzureAccessToken) in `cachedAzureAccessToken`. Record the `azureAccessTokenExpireTime` as
+         $t_0 + d\_{exp}$.
+6. Return $P$ as the additional KMS providers to [libmongocrypt](#libmongocrypt).
+
+<div id="obtaining-gcp-credentials">
+
+##### Obtaining GCP Credentials
+
+Set `HOST` to `metadata.google.internal`.
+
+Send an HTTP request to the URL `http://<HOST>/computeMetadata/v1/instance/service-accounts/default/token` with the
+header `Metadata-Flavor: Google`.
+
+If the HTTP response code is not 200, return an error including the body of the HTTP response in the error message.
+
+Parse the HTTP response body as JSON. If parsing fails, return an error including the body of the HTTP response in the
+error message.
+
+Check the parsed JSON for the field "access_token". If "access_token" is not present, return an error including the body
+of the HTTP response in the error message.
+
+Return "access_token" as the credential.
+
+##### Obtaining an Access Token for Azure Key Vault
+
+Virtual machines running on the Azure platform have one or more *Managed Identities* associated with them. From within
+the VM, an identity can be used by obtaining an access token via HTTP from the *Azure Instance Metadata Service* (IMDS).
+[See this documentation for more information](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http)
+
+..note:
+
+```
+To optimize for testability, it is recommended to implement an isolated
+abstraction for communication with IMDS. This will aide in the implementation
+of the prose tests of the communication with an IMDS server.
+```
+
+The below steps should be taken:
+
+01. Let $U$ be a new URL, initialized from the URL string `"http://169.254.169.254/metadata/identity/oauth2/token"`
+
+02. Add a query parameter `api-version=2018-02-01` to $U$.
+
+03. Add a query parameter `resource=https://vault.azure.net/` to $U$.
+
+04. Prepare an HTTP GET request $Req$ based on $U$.
+
+    <div class="note">
+
+    <div class="title">
+
+    Note
+
+    All query parameters on $U$ should be appropriately percent-encoded
```

Comment:
The `<div>` tags look unnecessary, and you probably want to use the consistent blockquote formatting with a note icon here.

Reply:
Done